torch/utils/data/datapipes/iter/grouping.py

```diff
@@ -205,6 +205,84 @@ def cmp_fn(a : Tuple[str, Any], b : Tuple[str, Any]):
+    def __init__(self,
+                 datapipe: IterDataPipe[T_co],
+                 group_key_fn: Callable,
+                 *,
+                 buffer_size: int = 10000,
+                 group_size: Optional[int] = None,
+                 unbatch_level: int = 0,
+                 guaranteed_group_size: Optional[int] = None,
+                 drop_remaining: bool = False):
+        if unbatch_level == 0:
+            self.datapipe = datapipe
+        else:
+            self.datapipe = datapipe.unbatch(unbatch_level=unbatch_level)
+        self.group_key_fn = group_key_fn
+        self.buffer_size = buffer_size
+        self.group_size = group_size
+        self.guaranteed_group_size = None
+        if group_size is not None:
+            self.guaranteed_group_size = group_size
+        if guaranteed_group_size is not None:
+            self.guaranteed_group_size = guaranteed_group_size
+        self.drop_remaining = drop_remaining
+
+    def __iter__(self):
+        buffer = {}
+        buffer_size = 0
+
+        for x in self.datapipe:
+            key = self.group_key_fn(x)
+            if buffer_size == self.buffer_size:
+                biggest_key = None
+                biggest_size = 0
+                for findkey in buffer.keys():
+                    if len(buffer[findkey]) > biggest_size:
+                        biggest_size = len(buffer[findkey])
+                        biggest_key = findkey
+
+                if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                    raise Exception('Failed to group items', str(buffer[biggest_key]))
+
+                if self.guaranteed_group_size is None or biggest_size >= self.guaranteed_group_size:
+                    yield buffer[biggest_key]
+
+                buffer_size -= len(buffer[biggest_key])
+                del buffer[biggest_key]
+
+            if key not in buffer:
+                buffer[key] = [x]
+            else:
+                buffer[key].append(x)
+            buffer_size += 1
+
+            if self.group_size is not None and self.group_size == len(buffer[key]):
+                yield buffer[key]
+                buffer_size -= len(buffer[key])
+                del buffer[key]
+
+        while buffer_size:
+            biggest_key = None
+            biggest_size = 0
+            for key in buffer.keys():
+                if len(buffer[key]) > biggest_size:
+                    biggest_size = len(buffer[key])
+                    biggest_key = key
+
+            if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                raise Exception('Failed to group items', str(buffer[biggest_key]))
+
+            if self.guaranteed_group_size is None or biggest_size >= self.guaranteed_group_size:
+                yield buffer[biggest_key]
```

Comment:
It seems we can have a function to return `biggest_key` to reduce the duplicated code.